### PR TITLE
Fixing twitter

### PR DESCRIPTION
**Pull request by zapu**
Use oembed API in `check_status`, the same that we use on clientside.

The difference is that we are still strict here on the proof text. Clientside uses the following regexp to match tweet text:
```
{ regex_capture: {
      , pattern: "^.*Verifying myself: I am ([A-Za-z0-9_]+) on (?:(?:https?://)?Keybase\\.io|<a href=\"https://t\\.co/[A-Za-z0-9_-]+\">https://t\\.co/[A-Za-z0-9_-]+</a>)[\\s\\p{Zs}]*\\. (\\S+) */.*$"
      , from: "tweet_contents_nw"
      , into: ["username_from_tweet_contents", "sig_from_tweet_contents"]
      , error: ["DELETED", "Could not find 'Verifying myself: I am %{username_keybase} on Keybase.io. %{sig_id_short}'"] } },
```

so it would accept any arbitrary text before the "Verifying myself...". This iced scraper can use proper HTML parser so even though the returned html is
```
<blockquote class=\"twitter-tweet\"><p lang=\"de\" dir=\"ltr\">Verifying myself: I am t_zapu on <a href=\"https://t.co/gaof5FoVSN\">https://t.co/gaof5FoVSN</a>. ...
```
we can properly parse it and ensure that tweet `<p>` starts matches `^Verifying myself...`.

### CODE REVIEW

**Comment by zapu** [comment]
Ok this seems to work, please review @mlsteele @maxtaco @malgorithms

File: src/scrapers/twitter.iced [hunk 1]
```diff
@@ -312,35 +312,78 @@ exports.TwitterScraper = class TwitterScraper extends BaseScraper
...
     new_api_url = urlmod.format u
-    @log "| rewrite #{api_url} -> #{new_api_url} (since twitter was going to redirect us anyways)"
-    api_url = new_api_url
+    @log "| use oembed API #{new_api_url} for tweet at #{api_url} (remote_id=#{remote_id})"
 
     # calls back with a v_code or null if it was ok
-    await @_get_url_body { url : api_url }, defer err, rc, html
+    await @_get_url_body { url : new_api_url }, defer err, rc, body
 
     if rc is v_codes.OK
```
Comment (zapu) [inline src/scrapers/twitter.iced:333]: there's a lot of logic here and a lot of early returns. I'm considering putting the whole thing inside of `try` block. Let me know if I should redo it in some other way.

**Review by malgorithms — APPROVED** [review]

File: src/scrapers/twitter.iced [hunk 2]
```diff
@@ -312,35 +312,84 @@ exports.TwitterScraper = class TwitterScraper extends BaseScraper
...
+      [_, author_username] = author_url_matches
+      if not(sncmp(username, author_username))
+        err = new Error("username from author_url didn't match, expected: #{username}, got: #{author_username}")
+        return cb err, v_codes.BAD_USERNAME
+
+      if typeof body_obj.html isnt 'string'
+        err = new Error("expected string for html, got #{typeof body_obj.html}")
+        return cb err, v_codes.CONTENT_MISSING
+
+      if body_obj.html.length > 1000
```
Comment (zapu) [inline src/scrapers/twitter.iced:382]: I found that very long stream of garbage like `"<p>" * 65535` can make `@libs.cheerio.load` hang for like a minute.

File: src/scrapers/base.iced [hunk 3]
```diff
@@ -105,8 +105,11 @@ class BaseScraper
     if rc is v_codes.OK
       try
         if opts.json
-          response.json().then (body) ->
+          response.json().then((body) ->
             cb err, rc, body
+          ).catch((err) ->
+            cb err, v_codes.BASE_HARD_ERROR
+          )
```
Comment (zapu) [inline src/scrapers/base.iced:112]: unhandled promise rejection bug when we ask for json but the site has returned something unparseable

**Comment by zapu** [comment]
@malgorithms spent some additional time exercising this thing and added some further hardening. none of these errors should be handled during normal operations, but rather when twitter malfunctions (or decides to actively mess with us)

File: src/constants.iced [hunk 4]
```diff
@@ -165,6 +165,8 @@ exports.constants = constants =
     EMPTY_JSON:        270
     BAD_API_URL:       280
     API_URL_MISMATCH:  281
+    UNEXPECTED_JSON:   282
+    CONTANT_MALFORMED: 283
```
Comment (mlsteele) [inline src/constants.iced]: These code set also appears in client protocol/avdl/keybase1/prove_common.avdl

File: src/scrapers/base.iced (hunk 3 above)
Comment (mlsteele) [reply src/scrapers/base.iced]: I don't think BASE_HARD_ERROR is meant to be used as a real code, rather it's the value that marks the start of the hard error section. How about CONTENT_FAILURE?

File: src/constants.iced (hunk 4 above)
Comment (zapu) [reply src/constants.iced]: crap. should i remove these new codes then?i thought it was internal only

**Review by mlsteele — APPROVED** [review]

**Comment by maxtaco** [comment]
You can use `schema3` if you'd like to make it easier to match JSON to a known schema.

Also, can you fix the minimist security warning by upgrading it?

**Comment by maxtaco** [comment]
Super!

**Pull request merged**